barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_format.cpp

```diff
@@ -17,12 +17,16 @@ template <typename Builder>
 void build_constraints(Builder& builder,
                        AcirFormat& constraint_system,
                        bool has_valid_witness_assignments,
-                       bool honk_recursion)
+                       bool honk_recursion,
+                       bool collect_gates_per_opcode)
 {
     constraint_system.gates_per_opcode = std::vector<size_t>(constraint_system.num_acir_opcodes);
```

Comment:
Nit: We could make this vector initialize conditional too. And this isn't very idiomatic, gates_per_opcode.resize(size) is

Reply:
If we don't resize the vector, we'd write zeroes past the vector length ): The only thing that is conditioned is the probably expensive call to `builder.get_num_gates();` at every step. 
I'll change the resizing method to the idiomatic one!

Reply:
Should I fully condition the writes to this array in all callsites instead? I don't think it's going to impact performance much and it'll make the code uglier, so I'm not sure

Reply:
Right, that is why I was wondering about the before and after time measurement

Reply:
<details>
<summary>
Measurements
</summary>

```
with accounting 

finished computation at Wed Jun 12 11:45:00 2024
elapsed time: 11.9586s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,


finished computation at Wed Jun 12 11:45:15 2024
elapsed time: 12.3645s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,

finished computation at Wed Jun 12 11:45:28 2024
elapsed time: 11.77s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,

finished computation at Wed Jun 12 11:45:41 2024
elapsed time: 10.859s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,


finished computation at Wed Jun 12 11:45:55 2024
elapsed time: 13.1604s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,



with accounting but disabled

finished computation at Wed Jun 12 11:36:00 2024
elapsed time: 10.6924s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,
        "gates_per_opcode": []
  }
]}%                                                                                                                                                                                                     
finished computation at Wed Jun 12 11:36:16 2024
elapsed time: 10.9079s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,
        "gates_per_opcode": []
  }
]}%                                                                                                                                                                                                     
finished computation at Wed Jun 12 11:36:30 2024
elapsed time: 11.2165s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,
        "gates_per_opcode": []
  }
]}%                                                                                                                                                                                                     
finished computation at Wed Jun 12 11:36:43 2024
elapsed time: 10.7484s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,
        "gates_per_opcode": []
  }
]}%                                                                                                                                                                                                     
finished computation at Wed Jun 12 11:36:57 2024
elapsed time: 11.1277s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712,
        "gates_per_opcode": []
  }
]}%                                                                                                                                                                                                     


without accounting

finished computation at Wed Jun 12 11:32:42 2024
elapsed time: 10.7138s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712
  }
]}%                                                                                                                                                                                                     
%  ~/aztec-packages/barretenberg/cpp/build/bin/bb gates -b ~/aztec-packages/noir-projects/noir-contracts/target/token_contract-Token-transfer.json
finished computation at Wed Jun 12 11:32:57 2024
elapsed time: 11.0694s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712
  }
]}%                                                                                                                                                                                                     
%  ~/aztec-packages/barretenberg/cpp/build/bin/bb gates -b ~/aztec-packages/noir-projects/noir-contracts/target/token_contract-Token-transfer.json
finished computation at Wed Jun 12 11:33:11 2024
elapsed time: 10.8229s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712
  }
]}%                                                                                                                                                                                                     
%  ~/aztec-packages/barretenberg/cpp/build/bin/bb gates -b ~/aztec-packages/noir-projects/noir-contracts/target/token_contract-Token-transfer.json
finished computation at Wed Jun 12 11:33:26 2024
elapsed time: 10.7055s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712
  }
]}%                                                                                                                                                                                                     
%  ~/aztec-packages/barretenberg/cpp/build/bin/bb gates -b ~/aztec-packages/noir-projects/noir-contracts/target/token_contract-Token-transfer.json
finished computation at Wed Jun 12 11:33:41 2024
elapsed time: 10.6757s
{"functions": [
  {
        "acir_opcodes": 148209,
        "circuit_size": 1949712
  }
]}%    
```

</details>

Reply:
There is a lot of variance, but clearly with enabled gate accounting the average is around a 5-10% performance hit, and with the  disabled accounting I'd say it's within run to run variance vs master